specs/impact-graph.md

```diff
@@ -0,0 +1,223 @@
+
+Unfortunately, measuring software impact metrics like carbon, water, energy is complex and very nuanced. 
+
+Modern applications are composed of many smaller pieces of software (components) running on many different environments, for example, private cloud, public cloud, bare-metal, virtualized, containerized, mobile, laptops, desktops, embedded, IoT. Many components that make up a typical software application are not run on resources you own or control, which makes including the impact of managed services in your measurement especially hard. 
+
+- How we model the environmental impacts of a complex application is through the creation of an **Impact Graph** which represents and models the thing we want to measure. 
+- Each **Impact Graph Node** in the graph represents one atomic thing we want to measure, it could be a software component, or part of the user joruney. 
+- To measure the environmental impact of one of the nodes in an **Impact Graph** we use a **Impant Mesurement Model**. 
+- There are many **Impact Measurement Models** out there and they all have different interfaces, some are APIs some are libraries, some are just CSV files. 
+- Since every **Impact Measurement Model** is bespoke to make them all interoperable we write **Impact Measurement Plugins** which adhere to a common **Impact Measurement Interface** specification.
+
+### Example
+
+Concrete example of how we might model the carbon impact of a web server using an **Impact Graph**.
+
+```
+My Web Server Impact Graph
+```
+
+We first create an **Impact Graph** that describes the ontology we want to measure. This might be an application like this case which is a Web Server, but it also might be a campaign or even an individual user journey. It could even be used to measure the embodied carbon of physical devices. What we are proposing here isn't limited to just software.
+
+```
+My Web Server Impact Graph
+├── Backend Server
+├── Load Balencer
+└── Cache Servers
+```
+
+We second break out the ontology into the nodes which we want to measure. The important thing is that we are breaking things out not as concepts, but as seperate and distint measurable components of the whole. 
+
+An **Impact Graph Node (IGN)** is something we want to measure distinct from every other node in the Impact Graph. That's it, we can clump all the servers together into one IGN regardless of what's running on them IF we want to measure them all together. We can break it up into one IGN per-process IF we want to measure per-process. The IGN describes an atomic thing we want to measure, that's it.
+
+What this means is that although a component might be "Backend Servers" since we'd use different approaches to measure "Intel Backend Servers" and "AMD Backend Servers" we might actually break it out into two IGNs, "AMD Backend Servers" and "Intel Backend Servers". This sounds complex but since our goal is to measure, measuring dictates how we strucutre the graph.

```

Comment:
We can add Cloud to the above details - measure "Intel Backend Servers" and "AMD Backend Servers"  or "Cloud Backend VMs"

Reply:
Oh my intention was Cloud. I think this is an interesting point would like to delve in with you and make sure we are on the same page.

We've discussed in previous ontology discussions and examples about a node being a single "software component" or at least that's how I felt each node in an ontology represented.

But as I've been thinking about this more and as we've been exploring this more at Intel a node in an ontology is not really a software component for us, it's a _thing_ we want to measure _separate_ from everything else we are measuring. 

So at Intel, we might want to work with a client and break a software component into the processes that run on AMD vs the processes that tun on Intel. So we can see how much energy the same process consumes when running on AMD chips vs Intel chips.

It's subtle but I think that's it's an important characteristic of the ontology, the lowest level of granularity isn't the process that's run, it's just a clustering of things we want to measure and that might not even relate to a process. Does that make sense?